File: test/warnings/array_types_warning/python/PackedArrayStructHasNoPackableFieldTest.py
import unittest
import os
import zserio

from testutils import getZserioApi, getApiDir
from ArrayTypesWarningTest import EXPECTED_WARNINGS

class PackedArrayStructHasNoPackableFieldTest(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.api = getZserioApi(__file__, "array_types_warning.zs",
                               expectedWarnings=EXPECTED_WARNINGS).packed_array_struct_has_no_packable_field

    def testWriteRead(self):
        packedArrayStructHasNoPackableField = self.api.PackedArrayStructHasNoPackableField(
            [
                self.api.StructWithPackable("A", 65),
                self.api.StructWithPackable("B", 66),
                self.api.StructWithPackable("C", 67)
            ],
            [
                self.api.UnionWithPackableField(),
                self.api.UnionWithPackableField(),
                self.api.UnionWithPackableField()
            ],
            [
                self.api.StructWithPackableArray("ABC", [65, 66, 67]),
                self.api.StructWithPackableArray("DEF", [68, 69, 70]),
                self.api.StructWithPackableArray("GHI", [71, 72, 73])
            ],
            [
                self.api.StructWithoutPackable(
                    4.0, zserio.BitBuffer(bytes([0xf0]), 5), 0, "A", [0, 0, 0], [True, False, True]
                ),
                self.api.StructWithoutPackable(
                    1.0, zserio.BitBuffer(bytes([0xe0]), 5), 0, "B", [0, 0, 0], [True, False, True]
                ),
                self.api.StructWithoutPackable(
                    0.0, zserio.BitBuffer(bytes([0xd0]), 5), 0, "C", [0, 0, 0], [True, False, True]
                )
            ],
            [
                self.api.EmptyStruct(),
                self.api.EmptyStruct(),
                self.api.EmptyStruct()
            ]
        )

        packedArrayStructHasNoPackableField.array2[0].field2 = self.api.TestEnum.ONE
        packedArrayStructHasNoPackableField.array2[1].field2 = self.api.TestEnum.TWO
        packedArrayStructHasNoPackableField.array2[2].field2 = self.api.TestEnum.ONE

        zserio.serialize_to_file(packedArrayStructHasNoPackableField, self.BLOB_NAME)
        readPackedArrayStructHasNoPackableField = zserio.deserialize_from_file(
            self.api.PackedArrayStructHasNoPackableField, self.BLOB_NAME
        )
        self.assertEqual(packedArrayStructHasNoPackableField, readPackedArrayStructHasNoPackableField)

    BLOB_NAME = os.path.join(getApiDir(os.path.dirname(__file__)),
                             "packed_array_struct_has_no_packable_field.blob")

File: test/warnings/array_types_warning/cpp/PackedArrayStructHasNoPackableFieldTest.cpp
#include "gtest/gtest.h"

#include "array_types_warning/packed_array_struct_has_no_packable_field/PackedArrayStructHasNoPackableField.h"

#include "zserio/SerializeUtil.h"

namespace array_types_warning
{
namespace packed_array_struct_has_no_packable_field
{

using allocator_type = PackedArrayStructHasNoPackableField::allocator_type;
template <typename T>
using vector_type = zserio::vector<T, allocator_type>;
using BitBuffer = zserio::BasicBitBuffer<allocator_type>;

class PackedArrayStructHasNoPackableFieldTest : public ::testing::Test
{
protected:
    static const std::string BLOB_NAME;
};

const std::string PackedArrayStructHasNoPackableFieldTest::BLOB_NAME =
        "warnings/array_types_warning/packed_array_struct_has_no_packable_field.blob";

TEST_F(PackedArrayStructHasNoPackableFieldTest, writeRead)
{
    PackedArrayStructHasNoPackableField packedArrayStructHasNoPackableField;

    packedArrayStructHasNoPackableField.setArray1({{
        StructWithPackable("A", 65),
        StructWithPackable("B", 66),
        StructWithPackable("C", 67)
    }});

    auto& array2 = packedArrayStructHasNoPackableField.getArray2();
    array2.resize(3);
    array2[0].setField2(TestEnum::ONE);
    array2[1].setField2(TestEnum::TWO);
    array2[2].setField2(TestEnum::ONE);

    packedArrayStructHasNoPackableField.setArray3({{
        StructWithPackableArray("ABC", {{ 65, 66, 67 }}),
        StructWithPackableArray("DEF", {{ 68, 69, 70 }}),
        StructWithPackableArray("GHI", {{ 71, 72, 73 }})
    }});

    packedArrayStructHasNoPackableField.setArray4({{
        StructWithoutPackable(4.0f, BitBuffer(vector_type<uint8_t>({0xF0}), 5), 0, "A",
                {{0, 0, 0}}, {{true, false, true}}),
        StructWithoutPackable(1.0f, BitBuffer(vector_type<uint8_t>({0xE0}), 5), 0, "B",
                {{0, 0, 0}}, {{true, false, true}}),
        StructWithoutPackable(0.0f, BitBuffer(vector_type<uint8_t>({0xD0}), 5), 0, "C",
                {{0, 0, 0}}, {{true, false, true}})
    }});

    packedArrayStructHasNoPackableField.getArray5().resize(3);

    zserio::serializeToFile(packedArrayStructHasNoPackableField, BLOB_NAME);
    auto readPackedArrayStructHasNoPackableField =
            zserio::deserializeFromFile<PackedArrayStructHasNoPackableField>(BLOB_NAME);
    ASSERT_EQ(packedArrayStructHasNoPackableField, readPackedArrayStructHasNoPackableField);
}

} // namespace packed_array_struct_has_no_packable_field
} // namespace array_types_warning
